wfe/wfe.go

```diff
@@ -1615,48 +1661,176 @@ func (wfe *WebFrontEndImpl) RevokeCert(
 	logEvent *requestEvent,
 	response http.ResponseWriter,
 	request *http.Request) {
-	body, _, prob := wfe.verifyPOST(ctx, logEvent, request, wfe.lookupJWK)
+
+	// The ACME specification handles the verification of revocation requests
+	// differently from other endpoints that always use one JWS authentication
+	// method. For this endpoint we need to accept a JWS with an embedded JWK, or
+	// a JWS with an embedded key ID, handling each case differently in terms of
+	// which certificates are authorized to be revoked by the requester
+
+	bodyBytes, err := ioutil.ReadAll(request.Body)
+	if err != nil {
+		wfe.sendError(
+			acme.InternalErrorProblem("unable to read request body"), response)
+		return
+	}
+	body := string(bodyBytes)
+
+	parsedJWS, err := wfe.parseJWS(body)
+	if err != nil {
+		wfe.sendError(
+			acme.MalformedProblem(err.Error()), response)
+		return
+	}
+
+	if prob := wfe.validPOST(request); prob != nil {
+		wfe.sendError(prob, response)
+		return
+	}
+
+	// Determine the authentication type for this request
+	authType, prob := checkJWSAuthType(parsedJWS)
 	if prob != nil {
 		wfe.sendError(prob, response)
 		return
 	}
 
+	// Handle the revocation request according to how it is authenticated, or if
+	// the authentication type is unknown, error immediately
+	if authType == embeddedKeyID {
+		prob = wfe.revokeCertByKeyID(ctx, logEvent, parsedJWS, request)
+	} else if authType == embeddedJWK {
+		prob = wfe.revokeCertByJWK(ctx, logEvent, parsedJWS, request)
+	} else {
+		prob = acme.MalformedProblem("Malformed JWS, no KeyID or embedded JWK")
+	}
+	if prob != nil {
+		wfe.sendError(prob, response)
+		return
+	}
+
+	response.WriteHeader(http.StatusOK)
+}
+
+func (wfe *WebFrontEndImpl) revokeCertByKeyID(
+	ctx context.Context,
+	logEvent *requestEvent,
+	jws *jose.JSONWebSignature,
+	request *http.Request) *acme.ProblemDetails {
+
+	pubKey, prob := wfe.lookupJWK(request, jws)
+	if prob != nil {
+		return prob
+	}
+
+	body, key, prob := wfe.verifyJWS(pubKey, jws, request)
+	if prob != nil {
+		return prob
+	}
+
+	keyID, err := keyToID(key)
+	if err != nil {
+		return acme.MalformedProblem(err.Error())
+	}
+
+	existingAcct := wfe.db.GetAccountByID(keyID)
+	if existingAcct == nil {
+		return acme.UnauthorizedProblem(fmt.Sprintf("Account with keyID %q does not exist", keyID))
+	}
+
+	// An account is only authorized to revoke its own certificates presently.
+	// TODO(@cpu): Allow an account to revoke another account's certificate if
+	// the revoker account has valid authorizations for all of the names in the
+	// to-be-revoked certificate.
+	authorizedToRevoke := func(cert *core.Certificate) *acme.ProblemDetails {
+		if cert.AccountID == existingAcct.ID {
+			return nil
+		}
+		return acme.UnauthorizedProblem(
+			fmt.Sprintf(
+				"Account %q did not request issuance of the to-be-revoked certificate",
```

Comment:
"The certificate being revoked is not associated with account %q". The "did not request issuance" is confusing because it sounds like Pebble is complaining that the revocation request isn't an issuance request.

Reply:
WFM: Fixed in 6fe0ac0